### PR TITLE
NON-160 Remove unused get NA's by booking ID and support filtering

### PR DESCRIPTION
**Pull request by Mjwillis**

### CODE REVIEW

**Review by andymarke — APPROVED** [review]
LGTM

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/BookingResourceIntTest.java [hunk 1]
```diff
@@ -889,28 +889,6 @@ public void getVisitsWithVisitorsFilteredByCancellationReason() {
...
-
-    @Test
-    public void getNonAssociationDetails_perpetrator_and_victim() {
-        final var response = testRestTemplate.exchange(
-            "/api/bookings/-2/non-association-details",
-            HttpMethod.GET,
-            createHttpEntity(authTokenHelper.getToken(AuthToken.NORMAL_USER), Map.of()),
-            String.class);
-
-        assertThatJsonFileAndStatus(response, 200, "offender_non_association_details_per_vic.json");
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/BookingResourceIntTest.java:911]: could you also delete any of these json files?

File: src/test/java/uk/gov/justice/hmpps/prison/service/OffenderNonAssociationsServiceTest.java [hunk 2]
```diff
@@ -1,231 +0,0 @@
...
-import java.util.List;
-import java.util.Optional;
-
-import static java.util.Collections.emptyList;
-import static org.assertj.core.api.Assertions.assertThat;
-import static org.assertj.core.api.Assertions.assertThatThrownBy;
-import static org.mockito.Mockito.when;
-
-@ExtendWith(MockitoExtension.class)
-public class OffenderNonAssociationsServiceTest {
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/prison/service/OffenderNonAssociationsServiceTest.java:30]: deleting this file doesn't seem right - sounds like it should be repurposed to call `retrieveByOffenderNo` instead

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/BookingResourceIntTest.java (hunk 1 above)
Comment (Mjwillis) [reply src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/BookingResourceIntTest.java:911]: yes!

File: src/test/java/uk/gov/justice/hmpps/prison/service/OffenderNonAssociationsServiceTest.java (hunk 2 above)
Comment (Mjwillis) [reply src/test/java/uk/gov/justice/hmpps/prison/service/OffenderNonAssociationsServiceTest.java:30]: yes

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderNonAssociation.java [hunk 3]
```diff
@@ -37,9 +37,13 @@ public class OffenderNonAssociation {
...
     private String agencyDescription;
 
+    @Schema(requiredMode = REQUIRED, description = "Prison ID", example = "PVI")
+    private String agencyId;
+
     @Schema(requiredMode = REQUIRED, description = "Description of living unit (e.g. cell) the offender is assigned to.", example = "PVI-1-2-4")
     private String assignedLivingUnitDescription;
 
-    @Schema(requiredMode = REQUIRED, description = "Id of living unit (e.g. cell) the offender is assigned to.", example = "123")
+    @Deprecated
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderNonAssociation.java:46]: This is being removed, no code in github seems to use it for NAs

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderNonAssociationDetail.java [hunk 4]
```diff
@@ -33,11 +34,11 @@ public class OffenderNonAssociationDetail {
     @Schema(requiredMode = REQUIRED, description = "The non-association type description", example = "Do Not Locate on Same Wing")
     private String typeDescription;
 
-    @Schema(requiredMode = REQUIRED, description = "Date and time the mom-association is effective from. In Europe/London (ISO 8601) format without timezone offset e.g. YYYY-MM-DDTHH:MM:SS.", example = "2019-12-01T13:34:00")
-    private LocalDateTime effectiveDate;
+    @Schema(requiredMode = REQUIRED, description = "Date and time the mom-association is effective from", example = "2019-12-01")
+    private LocalDate effectiveDate;
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderNonAssociationDetail.java:38]: These are dates not date times

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderNonAssociation.java [hunk 5]
```diff
@@ -37,9 +37,13 @@ public class OffenderNonAssociation {
     @Schema(requiredMode = REQUIRED, description = "Description of the agency (e.g. prison) the offender is assigned to.", example = "Pentonville (PVI)")
     private String agencyDescription;
 
+    @Schema(requiredMode = REQUIRED, description = "Prison ID", example = "PVI")
```
[the remark this answers is not in the document]
Comment (petergphillips) [reply src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderNonAssociation.java:40]: new field that isn't tested?

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderNonAssociationDetails.java [hunk 6]
```diff
@@ -34,6 +34,8 @@ public class OffenderNonAssociationDetails {
     @Schema(requiredMode = REQUIRED, description = "Description of the agency (e.g. prison) the offender is assigned to.", example = "Moorland (HMP & YOI)")
     private String agencyDescription;
 
+    @Schema(requiredMode = REQUIRED, description = "Prison ID", example = "MDI")
+    private String agencyId;
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderNonAssociationDetails.java:38]: new field that isn't tested?

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java [hunk 7]
```diff
@@ -669,9 +669,11 @@ public OffenderContacts getOffenderContacts(
     @Deprecated
     @SlowReportQuery
     public OffenderNonAssociationDetails getNonAssociationDetails(
-        @Parameter(name = "offenderNo", description = "Offender No", example = "A1234AA", required = true) @PathVariable(value = "offenderNo") @NotNull final String offenderNo) {
+        @Parameter(name = "offenderNo", description = "Offender No", example = "A1234AA", required = true) @PathVariable(value = "offenderNo") @NotNull final String offenderNo,
+        @RequestParam(value = "currentPrisonOnly", required = false, defaultValue = "false") @Parameter(description = "Returns only non-association details for this prisoner in the same prison") final Boolean currentPrisonOnly,
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderResource.java:673]: two new params that need testing?

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderNonAssociationsService.java [hunk 8]
```diff
@@ -30,32 +32,24 @@ public class OffenderNonAssociationsService {
...
         log.debug("Fetching non-associations for offender no '{}'", offenderNo);
+        final var now = LocalDate.now(clock);
 
+        final var latestBooking = bookingRepository.findWithDetailsByOffenderNomsIdAndBookingSequence(offenderNo, 1)
+            .orElseThrow(EntityNotFoundException.withMessage("Booking not found for offender %s.", offenderNo));
         final List<OffenderNonAssociationDetail> nonAssociations
-            = offenderNonAssociationDetailRepository.findAllByOffenderBooking_Offender_NomsIdOrderByEffectiveDateAsc(offenderNo);
+            = offenderNonAssociationDetailRepository.findAllByOffenderBooking_Offender_NomsIdOrderByEffectiveDateAsc(offenderNo)
+                .stream().filter(
+                    na -> !excludeInactive || ((na.getExpiryDate() == null || !na.getExpiryDate().isBefore(now)) && !na.getEffectiveDate().isAfter(now)))
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/service/OffenderNonAssociationsService.java:47]: new filters that would be good to test

**Review by petergphillips — APPROVED** [review]
LGTM

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderNonAssociationsService.java (hunk 8 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/service/OffenderNonAssociationsService.java:47]: tested in service mock layer

**Pull request merged**